File: ShowcaseView_java/library/src/main/java/com/smorodov/showcaseview/anim/AnimationUtils.java
package com.smorodov.showcaseview.anim;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.util.Log;
import android.view.View;

public class AnimationUtils {

    public static final int DEFAULT_DURATION = 300;

    private static final String ALPHA = "alpha";
    private static final float INVISIBLE = 0f;
    private static final float VISIBLE = 1f;
    private static final String COORD_X = "x";
    private static final String COORD_Y = "y";
    private static final int INSTANT = 0;

    public static float getX(View view) {
        return view.getX();
    }

    public static float getY(View view) {
        return view.getY();
    }

    public static void hide(View view) {
        view.setAlpha(INVISIBLE);
    }

    public static ObjectAnimator createFadeInAnimation(Object target, final AnimationStartListener listener) {
        return createFadeInAnimation(target, DEFAULT_DURATION, listener);
    }

    public static ObjectAnimator createFadeInAnimation(Object target, int duration, final AnimationStartListener listener) {
        ObjectAnimator oa = ObjectAnimator.ofFloat((View) target, View.ALPHA, VISIBLE);
        oa.setDuration(duration).addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {
                listener.onAnimationStart();
            }

            @Override
            public void onAnimationEnd(Animator animator) {
            }

            @Override
            public void onAnimationCancel(Animator animator) {
            }

            @Override
            public void onAnimationRepeat(Animator animator) {
            }
        });
        return oa;
    }

    public static ObjectAnimator createFadeOutAnimation(Object target, final AnimationEndListener listener) {
        return createFadeOutAnimation(target, DEFAULT_DURATION, listener);
    }

    public static ObjectAnimator createFadeOutAnimation(Object target, int duration, final AnimationEndListener listener) {
        ObjectAnimator oa = ObjectAnimator.ofFloat((View) target, View.ALPHA, INVISIBLE);
        oa.setDuration(duration).addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {
            }

            @Override
            public void onAnimationEnd(Animator animator) {
                listener.onAnimationEnd();
            }

            @Override
            public void onAnimationCancel(Animator animator) {
            }

            @Override
            public void onAnimationRepeat(Animator animator) {
            }
        });
        return oa;
    }

    public static AnimatorSet createMovementAnimation(View view, float canvasX, float canvasY,
                                                      float offsetStartX, float offsetStartY,
                                                      float offsetEndX, float offsetEndY, final AnimationEndListener listener) {
        view.setAlpha(INVISIBLE);
        ObjectAnimator alphaIn = ObjectAnimator.ofFloat(view, ALPHA, INVISIBLE, VISIBLE).setDuration(500);
        ObjectAnimator setUpX = ObjectAnimator.ofFloat(view, COORD_X, canvasX + offsetStartX).setDuration(INSTANT);
        ObjectAnimator setUpY = ObjectAnimator.ofFloat(view, COORD_Y, canvasY + offsetStartY).setDuration(INSTANT);
        ObjectAnimator moveX = ObjectAnimator.ofFloat(view, COORD_X, canvasX + offsetEndX).setDuration(1000);
        ObjectAnimator moveY = ObjectAnimator.ofFloat(view, COORD_Y, canvasY + offsetEndY).setDuration(1000);
        moveX.setStartDelay(1000);
        moveY.setStartDelay(1000);

        ObjectAnimator alphaOut = ObjectAnimator.ofFloat(view, ALPHA, INVISIBLE).setDuration(500);
        alphaOut.setStartDelay(2500);

        AnimatorSet as = new AnimatorSet();
        as.play(setUpX).with(setUpY).before(alphaIn).before(moveX).with(moveY).before(alphaOut);

        as.addListener(new Animator.AnimatorListener()
                       {

                           @Override
                           public void onAnimationStart(Animator animation) {
                               Log.i("smorodov","animation start");
                           }

                           @Override
                           public void onAnimationEnd(Animator animation) {
                               Log.i("smorodov","animation end");
                               listener.onAnimationEnd();
                           }

                           @Override
                           public void onAnimationCancel(Animator animation) {

                           }

                           @Override
                           public void onAnimationRepeat(Animator animation) {

                           }
                       }
        );


        return as;
    }

    public static AnimatorSet createPointingAnimation(View view, float x, float y, final AnimationEndListener listener) {
        ObjectAnimator alphaIn = ObjectAnimator.ofFloat(view, ALPHA, INVISIBLE, VISIBLE).setDuration(1000);
        ObjectAnimator alphaOut = ObjectAnimator.ofFloat(view, ALPHA, VISIBLE, INVISIBLE).setDuration(1000);
        ObjectAnimator setUpX = ObjectAnimator.ofFloat(view, COORD_X, x).setDuration(INSTANT);
        ObjectAnimator setUpY = ObjectAnimator.ofFloat(view, COORD_Y, y).setDuration(INSTANT);

        AnimatorSet as = new AnimatorSet();
        as.addListener(new Animator.AnimatorListener()
                       {

                           @Override
                           public void onAnimationStart(Animator animation) {
                               Log.i("smorodov","animation start");
                           }

                           @Override
                           public void onAnimationEnd(Animator animation) {
                               Log.i("smorodov","animation end");
                               listener.onAnimationEnd();
                           }

                           @Override
                           public void onAnimationCancel(Animator animation) {
                           }

                           @Override
                           public void onAnimationRepeat(Animator animation) {

                           }
                       }
        );

        as.play(setUpX).with(setUpY);
        as.play(alphaIn);
        as.play(alphaOut).after(2000);
        return as;
    }

    public interface AnimationStartListener {
        void onAnimationStart();
    }

    public interface AnimationEndListener {
        void onAnimationEnd();
    }
}

File: ShowcaseView_kotlin/library/src/main/java/com/smorodov/showcaseview/anim/AnimationUtils.kt
package com.smorodov.showcaseview.anim

import android.animation.Animator
import android.animation.Animator.AnimatorListener
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.util.Log
import android.view.View

object AnimationUtils {
    const val DEFAULT_DURATION = 300
    private const val ALPHA = "alpha"
    private const val INVISIBLE = 0f
    private const val VISIBLE = 1f
    private const val COORD_X = "x"
    private const val COORD_Y = "y"
    private const val INSTANT = 0
    fun getX(view: View): Float {
        return view.x
    }

    fun getY(view: View): Float {
        return view.y
    }

    fun hide(view: View?) {
        view!!.alpha = INVISIBLE
    }

    fun createFadeInAnimation(target: Any?, listener: AnimationStartListener): ObjectAnimator {
        return createFadeInAnimation(target, DEFAULT_DURATION, listener)
    }

    fun createFadeInAnimation(
        target: Any?,
        duration: Int,
        listener: AnimationStartListener
    ): ObjectAnimator {
        val oa = ObjectAnimator.ofFloat(target as View?, View.ALPHA, VISIBLE)
        oa.setDuration(duration.toLong()).addListener(object : AnimatorListener {
            override fun onAnimationStart(animator: Animator) {
                listener.onAnimationStart()
            }

            override fun onAnimationEnd(animator: Animator) {}
            override fun onAnimationCancel(animator: Animator) {}
            override fun onAnimationRepeat(animator: Animator) {}
        })
        return oa
    }

    fun createFadeOutAnimation(target: Any?, listener: AnimationEndListener): ObjectAnimator {
        return createFadeOutAnimation(target, DEFAULT_DURATION, listener)
    }

    fun createFadeOutAnimation(
        target: Any?,
        duration: Int,
        listener: AnimationEndListener
    ): ObjectAnimator {
        val oa = ObjectAnimator.ofFloat(target as View?, View.ALPHA, INVISIBLE)
        oa.setDuration(duration.toLong()).addListener(object : AnimatorListener {
            override fun onAnimationStart(animator: Animator) {}
            override fun onAnimationEnd(animator: Animator) {
                listener.onAnimationEnd()
            }

            override fun onAnimationCancel(animator: Animator) {}
            override fun onAnimationRepeat(animator: Animator) {}
        })
        return oa
    }

    fun createMovementAnimation(
        view: View?, canvasX: Float, canvasY: Float,
        offsetStartX: Float, offsetStartY: Float,
        offsetEndX: Float, offsetEndY: Float, listener: AnimationEndListener
    ): AnimatorSet {
        view!!.alpha = INVISIBLE
        val alphaIn = ObjectAnimator.ofFloat(view, ALPHA, INVISIBLE, VISIBLE).setDuration(500)
        val setUpX = ObjectAnimator.ofFloat(view, COORD_X, canvasX + offsetStartX).setDuration(
            INSTANT.toLong()
        )
        val setUpY = ObjectAnimator.ofFloat(view, COORD_Y, canvasY + offsetStartY).setDuration(
            INSTANT.toLong()
        )
        val moveX = ObjectAnimator.ofFloat(view, COORD_X, canvasX + offsetEndX).setDuration(1000)
        val moveY = ObjectAnimator.ofFloat(view, COORD_Y, canvasY + offsetEndY).setDuration(1000)
        moveX.startDelay = 1000
        moveY.startDelay = 1000
        val alphaOut = ObjectAnimator.ofFloat(view, ALPHA, INVISIBLE).setDuration(500)
        alphaOut.startDelay = 2500
        val `as` = AnimatorSet()
        `as`.play(setUpX).with(setUpY).before(alphaIn).before(moveX).with(moveY).before(alphaOut)
        `as`.addListener(object : AnimatorListener {
            override fun onAnimationStart(animation: Animator) {
                Log.i("smorodov", "animation start")
            }

            override fun onAnimationEnd(animation: Animator) {
                Log.i("smorodov", "animation end")
                listener.onAnimationEnd()
            }

            override fun onAnimationCancel(animation: Animator) {}
            override fun onAnimationRepeat(animation: Animator) {}
        }
        )
        return `as`
    }

    fun createPointingAnimation(
        view: View?,
        x: Float,
        y: Float,
        listener: AnimationEndListener
    ): AnimatorSet {
        val alphaIn = ObjectAnimator.ofFloat(view, ALPHA, INVISIBLE, VISIBLE).setDuration(1000)
        val alphaOut = ObjectAnimator.ofFloat(view, ALPHA, VISIBLE, INVISIBLE).setDuration(1000)
        val setUpX = ObjectAnimator.ofFloat(view, COORD_X, x).setDuration(INSTANT.toLong())
        val setUpY = ObjectAnimator.ofFloat(view, COORD_Y, y).setDuration(INSTANT.toLong())
        val `as` = AnimatorSet()
        `as`.addListener(object : AnimatorListener {
            override fun onAnimationStart(animation: Animator) {
                Log.i("smorodov", "animation start")
            }

            override fun onAnimationEnd(animation: Animator) {
                Log.i("smorodov", "animation end")
                listener.onAnimationEnd()
            }

            override fun onAnimationCancel(animation: Animator) {}
            override fun onAnimationRepeat(animation: Animator) {}
        }
        )
        `as`.play(setUpX).with(setUpY)
        `as`.play(alphaIn)
        `as`.play(alphaOut).after(2000)
        return `as`
    }

    fun interface AnimationStartListener {
        fun onAnimationStart()
    }

    fun interface AnimationEndListener {
        fun onAnimationEnd()
    }
}